### PR TITLE
Ux 225/implement tls reloader

### PR DESCRIPTION
**Pull request by sago2k8**
Implement hot reload for TLS certificates, this is meant to be used by clients.

### CODE REVIEW

File: backend/pkg/tls/ca.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	digest := sha256.Sum256(ca)
+	if digest == r.digest {
+		r.logger.Debug("CA cert did not change", zap.String("path", r.caPath))
+		return nil
+	}
+	cfg := r.tc.Load().Clone()
+	if cfg.ClientCAs == nil {
+		cfg.ClientCAs = x509.NewCertPool()
+	}
+	if !cfg.ClientCAs.AppendCertsFromPEM(ca) {
```
Comment (andresaristizabal) [inline backend/pkg/tls/ca.go:120]: I understand this fn add the new ca to the existing ones, do we remove the old ones at some point?

File: backend/pkg/tls/ca.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	ticker := time.NewTicker(refreshInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-quitCtx.Done():
+			return
+		case <-ticker.C:
+			if err := r.load(); err != nil {
+				r.logger.Error("error while reloading the CA", zap.Any("err", err))
```
Comment (andresaristizabal) [inline backend/pkg/tls/ca.go:100]: Do you think we should include we are going to try again in x minutes? in order to user try to fix and it will try again automatically

File: backend/pkg/tls/keypair.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+	"github.com/twmb/tlscfg"
+	"go.uber.org/zap"
+)
+
+// MaybeWithDynamicDiskKeyPair returns a function that configures a certificate
+// on the tls Config that is read from a pair of files on disk and updated
+// whenever the content of those files changes.
+//
+// Files are polled every 5 minutes for modifications.
```
Comment (andresaristizabal) [inline backend/pkg/tls/keypair.go:30]: [nit]: I think we should use the config name in TLS instead of say 5 minutes

File: backend/pkg/tls/keypair_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/stretchr/testify/require"
+	"github.com/twmb/tlscfg"
+)
+
+func TestMaybeWithDynamicDiskKeyPair(t *testing.T) {
+	// Setup test environment
+	testSetup := setupTLSTest(t)
+	defer os.RemoveAll(testSetup.TmpDir)
+
+	t.Run("empty paths return nil", func(t *testing.T) {
```
Comment (andresaristizabal) [inline backend/pkg/tls/keypair_test.go:30]: I think we can add a test for ca path and key empty path, no?

File: backend/pkg/tls/ca.go (hunk 1 above)
Comment (sago2k8) [reply backend/pkg/tls/ca.go:120]: We don't, this is how Serverless does it and it makes sense imho. Certificates will be rotated by kubernetes because they will be expired soon, if the reason for rotation is a certificate breach I would expect the pod to be restarted.

File: backend/pkg/tls/tls_test_utils.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	err := os.WriteFile(path, certPEM, 0o600)
+	require.NoError(t, err, "Failed to write certificate to file")
+}
+
+// writeCertToPEMFile writes a certificate to a file in PEM format
+func writeCertToPEMFile(t *testing.T, path string, cert *x509.Certificate) {
+	certPEM := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw})
+	err := os.WriteFile(path, certPEM, 0o600)
+	require.NoError(t, err, "Failed to write certificate to PEM file")
+}
```
Comment (malinskibeniamin) [inline backend/pkg/tls/tls_test_utils.go:141]: Why do we have 2 of these functions? They are exactly the same.

Comment (sago2k8) [reply backend/pkg/tls/tls_test_utils.go:141]: Good call. thanks

**Review by weeco — APPROVED** [review]
